File: src/lib/golingvu/base_value.go
package golingvu

// BaseValue extends BaseSetup for TDT pattern
type BaseValue struct {
	*BaseSetup
	TableRows [][]interface{}
}

// NewBaseValue creates a new BaseValue instance
func NewBaseValue(
	features []string,
	tableRows [][]interface{},
	confirmCB interface{},
	initialValues interface{},
) *BaseValue {
	// For TDT, actions will be Should and checks will be Expected
	// We'll process them differently in setup
	baseSetup := NewBaseSetup(features, []interface{}{}, []interface{}{}, confirmCB, initialValues)
	return &BaseValue{
		BaseSetup: baseSetup,
		TableRows: tableRows,
	}
}

// SetupThat implements BaseSetup's abstract method
func (bv *BaseValue) SetupThat(
	subject interface{},
	testResourceConfiguration ITTestResourceConfiguration,
	artifactory func(string, interface{}),
	setupCB interface{},
	initialValues interface{},
) (interface{}, error) {
	// This should be implemented by the adapter
	return subject, nil
}

// Value is an alias for Setup for TDT pattern
func (bv *BaseValue) Value(
	subject interface{},
	key string,
	testResourceConfiguration ITTestResourceConfiguration,
	tester func(interface{}) bool,
	artifactory func(string, interface{}),
	suiteNdx *int,
) (interface{}, error) {
	return bv.Setup(subject, key, testResourceConfiguration, tester, artifactory, suiteNdx)
}

File: src/lib/rubeno/lib/base_value.rb
module Rubeno
  # BaseValue extends BaseSetup for TDT pattern.
  # Sets up table data for table-driven testing.
  class BaseValue < BaseSetup
    attr_accessor :table_rows, :adapter
    
    def initialize(features, table_rows, confirm_cb, initial_values)
      # For TDT, actions will be Should and checks will be Expected
      super(features, [], [], confirm_cb, initial_values)
      @table_rows = table_rows || []
      @adapter = nil
    end
    
    def setup_that(subject, test_resource_configuration, artifactory, setup_cb, initial_values)
      # This should be implemented by subclasses
      raise NotImplementedError, "Subclasses must implement setup_that"
    end
    
    def after_each(store, key, artifactory)
      if @adapter && @adapter.respond_to?(:after_each)
        @adapter.after_each(store, key, artifactory)
      else
        store
      end
    end
    
    # Alias for TDT pattern
    def value(subject, key, test_resource_configuration, tester, artifactory = nil, suite_ndx = nil)
      setup(subject, key, test_resource_configuration, tester, artifactory, suite_ndx)
    end
  end
end
